### ISSUE TITLE
[oauth] Can a binding set the tokenExpiresInSeconds buffer?

### ISSUE THREAD

**Issue by mhilbush** [openhab/openhab-core#1438]
Over the past few months, I've seen a couple occurrences where my Ecobee binding gets a token expired response from the Ecobee API. The binding polls the Ecobee API every 30 or so seconds. Therefore, for this to happen only a couple times over such a long period of time would suggest some type of rare edge condition.

I can think of one scenario that might cause this to occur.

By default, the OAuth feature sets `tokenExpiresInSeconds` to 10 seconds by default. This is the buffer, or wiggle room, that's supposed to ensure that a token won't expire shortly after it is returned from the call to `getAccessTokenResponse`. Basically, it says that if a token expires in 3600 seconds (e.g. this is what Ecobee uses), then the OAuth Client Service will use 3590 seconds as the expiration time.

I've seen cases where calls to the Ecobee API can take longer than 10 seconds. Calls of this duration are rare, but they do happen from time to time. If one of these calls happens to occur just before the 3590 second expiry time, it's conceivable that the Ecobee API will receive a request containing a token that has just expired.

I've also seen where the Spotify binding [retries its call to the spotify API](https://github.com/openhab/openhab-addons/blob/127f05d7cbb865ae2f7a939c761707f0479d5143/bundles/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/api/SpotifyApi.java#L284) if it gets a token expired response from the API. This would seem to add credibility to the need to have a little more wiggle room between the call to `getAccessTokenResponse` and the call to the service's API.

I was hoping to set the `tokenExpiredInSeconds` value to something greater than the default of 10 seconds. The `OAuthClientService` has a [setTokenExpiresInBuffer](https://github.com/openhab/openhab-core/blob/3811cb95604f69f6a3d77da2caf391d81b05458e/bundles/org.openhab.core.auth.oauth2client/src/main/java/org/openhab/core/auth/oauth2client/internal/OAuthClientServiceImpl.java#L357) API, but it's not exposed to bindings.

Can anyone think of a reason why this method is not exposed? It seems like a reasonable thing that a binding might want to alter.

**Comment by mhilbush** [comment]
I now see that I can use `AccessTokenResponse.isExpired()` to check expiration status using my own `tokenExpiredInBuffer`, then force a refresh of the token using `OAuthClientService.refreshToken()`. That's exactly what I wanted.

**Comment by splatch** [comment]
@mhilbush does answer from oauth service indicate lifetime of a token? I believe that token refresh should take place on the lifetime basis and not static interval.

**Comment by mhilbush** [comment]
@splatch Yes, the `AccessTokenResponse` includes the lifetime of the token. The `OAuthClientService` will refresh the token in `lifetime - tokenExpiredInBuffer` seconds. The problem is that the default `tokenExpiredInBuffer` is 10 seconds, which may be too short in some cases. As there's no way to change from the default value of `tokenExpiredInBuffer`, the only solution is for the binding to do the token refresh every `lifetime - myCustomTokenExpiredInBuffer` seconds.

I opened this issue because I don't understand why the binding is not allowed to change `tokenExpiredInBuffer` from the default of 10 seconds. There's an internal API to do it, but that API is not exposed to the binding.

**Comment by cweitkamp** [comment]
I think it is okay to set an individual value for `tokenExpiresInSeconds` as long as it is unique for one binding / one `OAuthClientService` instance. It should not affect others. We can expose the setter.

The code seems a little bit weird. The `OAuthFactoryImpl` has a [setTokenExpiresInBuffer](https://github.com/openhab/openhab-core/blob/240c245b163b91c919925b9855c3657fa2bd430f/bundles/org.openhab.core.auth.oauth2client/src/main/java/org/openhab/core/auth/oauth2client/internal/OAuthFactoryImpl.java#L138) too which also is not exposed. Additionally the factory passes its value via a `PersistedParams` objects to a new client - which is okay. For restoring a previous available client (e.g. on restart) the `tokenExpiresInSeconds` from the factory is used for it. Even if the client has its own value in the persisted data. IMO this is a bug.

The [setTokenExpiresInBuffer](https://github.com/openhab/openhab-core/blob/3811cb95604f69f6a3d77da2caf391d81b05458e/bundles/org.openhab.core.auth.oauth2client/src/main/java/org/openhab/core/auth/oauth2client/internal/OAuthClientServiceImpl.java#L357) method of the client updates the value in the clients `persistedParams` property which on the one hand never gets saved again and on the other hand is not taken into account later. A second bug ...

**Comment by mhilbush** [comment]
You know, I'm glad you looked at this. When I looked at it I was scratching my head trying to figure out how a new `tokenExpiresInBuffer` would get persisted if the `setTokenExpiresInBuffer` method was exposed to the binding.

I also couldn't figure out why OAuthClientServiceImpl` and `OAuthFactoryImpl` both had a `setTokenExpiresInBuffer` method. The one on the factory seemed unnecessary to me, but I figured it was just something I didn't understand.

**Comment by cweitkamp** [comment]
Yes, the method seems not to be needed inside the factory. Will you submit a PR for it or shall I propose a draft?

**Comment by mhilbush** [comment]
I'm willing to take a stab at a PR. But I may need a little guidance along the way.

**Comment by cweitkamp** [comment]
Cool. Do not hesitate to ask any questions.

**Comment by J-N-K** [comment]
@mhilbush Are you still working on this?

**Comment by mhilbush** [comment]
No, I never got around to coming up with a fix.
